File: parser/summary_period.go
package parser

import (
	"regexp"
	"strconv"
	"strings"
	"time"
)

type MonthlySummaryPeriod struct {
	Year  int
	Month time.Month
}

var (
	summaryRelativePreviousPattern = regexp.MustCompile(`\bthang\s+(?:truoc|vua\s+roi)\b`)
	summaryCurrentPattern          = regexp.MustCompile(`\bthang\s+nay\b`)
	summaryMonthYearPattern        = regexp.MustCompile(`\b(0?[1-9]|1[0-2])\s*/\s*((?:19|20)\d{2})\b`)
	summaryYearMonthPattern        = regexp.MustCompile(`\b((?:19|20)\d{2})\s*(?:/|\s)\s*(0?[1-9]|1[0-2])\b`)
	summaryNamedMonthPattern       = regexp.MustCompile(`\bthang\s+(0?[1-9]|1[0-2])(?:\s*/?\s*((?:19|20)\d{2}))?\b`)
	summaryBareMonthPattern        = regexp.MustCompile(`^(0?[1-9]|1[0-2])(?:\s*/?\s*((?:19|20)\d{2}))?$`)
)

func ParseMonthlySummaryPeriod(input string, now time.Time) (MonthlySummaryPeriod, bool) {
	normalized := normalizeForIntent(input)
	current := MonthlySummaryPeriod{Year: now.Year(), Month: now.Month()}
	if normalized == "" {
		return current, true
	}
	if strings.HasPrefix(normalized, "/summary") {
		fields := strings.Fields(normalized)
		if len(fields) <= 1 {
			return current, true
		}
		normalized = strings.Join(fields[1:], " ")
	}
	if normalized == "" || summaryCurrentPattern.MatchString(normalized) {
		return current, true
	}
	if summaryRelativePreviousPattern.MatchString(normalized) {
		firstOfCurrentMonth := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, now.Location())
		previous := firstOfCurrentMonth.AddDate(0, -1, 0)
		return MonthlySummaryPeriod{Year: previous.Year(), Month: previous.Month()}, true
	}
	if match := summaryMonthYearPattern.FindStringSubmatch(normalized); match != nil {
		return periodFromParts(match[2], match[1])
	}
	if match := summaryYearMonthPattern.FindStringSubmatch(normalized); match != nil {
		return periodFromParts(match[1], match[2])
	}
	if match := summaryNamedMonthPattern.FindStringSubmatch(normalized); match != nil {
		year := strconv.Itoa(now.Year())
		if match[2] != "" {
			year = match[2]
		}
		return periodFromParts(year, match[1])
	}
	if match := summaryBareMonthPattern.FindStringSubmatch(normalized); match != nil {
		year := strconv.Itoa(now.Year())
		if match[2] != "" {
			year = match[2]
		}
		return periodFromParts(year, match[1])
	}
	return MonthlySummaryPeriod{}, false
}

func periodFromParts(yearText, monthText string) (MonthlySummaryPeriod, bool) {
	year, err := strconv.Atoi(yearText)
	if err != nil {
		return MonthlySummaryPeriod{}, false
	}
	month, err := strconv.Atoi(monthText)
	if err != nil || month < 1 || month > 12 {
		return MonthlySummaryPeriod{}, false
	}
	return MonthlySummaryPeriod{Year: year, Month: time.Month(month)}, true
}

File: src/parser/summary_period.ts
import { normalizeForIntent } from './intent.ts';

export interface MonthlySummaryPeriod {
  year: number;
  month: number;
}

const PREVIOUS_PATTERN = /\bthang\s+(?:truoc|vua\s+roi)\b/u;
const CURRENT_PATTERN = /\bthang\s+nay\b/u;
const MONTH_YEAR_PATTERN = /\b(0?[1-9]|1[0-2])\s*\/\s*((?:19|20)\d{2})\b/u;
const YEAR_MONTH_PATTERN = /\b((?:19|20)\d{2})\s*(?:\/|\s|-)\s*(0?[1-9]|1[0-2])\b/u;
const NAMED_MONTH_PATTERN = /\bthang\s+(0?[1-9]|1[0-2])(?:\s*\/?\s*((?:19|20)\d{2}))?\b/u;
const BARE_MONTH_PATTERN = /^(0?[1-9]|1[0-2])(?:\s*\/?\s*((?:19|20)\d{2}))?$/u;

export function parseMonthlySummaryPeriod(
  input: string,
  now: Date,
  timeZone = 'Asia/Ho_Chi_Minh',
): MonthlySummaryPeriod | undefined {
  const normalized = normalizeForIntent(input);
  const current = currentPeriod(now, timeZone);
  if (!normalized) return current;

  let value = normalized;
  if (value.startsWith('/summary')) {
    value = value.slice('/summary'.length).trim();
    if (!value) return current;
  }
  if (!value || CURRENT_PATTERN.test(value)) return current;
  if (PREVIOUS_PATTERN.test(value)) return previousPeriod(current);

  const monthYear = MONTH_YEAR_PATTERN.exec(value);
  if (monthYear) return periodFromParts(monthYear[2]!, monthYear[1]!);
  const yearMonth = YEAR_MONTH_PATTERN.exec(value);
  if (yearMonth) return periodFromParts(yearMonth[1]!, yearMonth[2]!);
  const named = NAMED_MONTH_PATTERN.exec(value);
  if (named) return periodFromParts(named[2] ?? String(current.year), named[1]!);
  const bare = BARE_MONTH_PATTERN.exec(value);
  if (bare) return periodFromParts(bare[2] ?? String(current.year), bare[1]!);
  return undefined;
}

function currentPeriod(now: Date, timeZone: string): MonthlySummaryPeriod {
  const parts = new Intl.DateTimeFormat('en-CA', { timeZone, year: 'numeric', month: 'numeric' })
    .formatToParts(now);
  const year = Number(parts.find((part) => part.type === 'year')?.value);
  const month = Number(parts.find((part) => part.type === 'month')?.value);
  if (!Number.isInteger(year) || !Number.isInteger(month)) throw new Error('invalid clock date');
  return { year, month };
}

function previousPeriod(current: MonthlySummaryPeriod): MonthlySummaryPeriod {
  return current.month === 1
    ? { year: current.year - 1, month: 12 }
    : { year: current.year, month: current.month - 1 };
}

function periodFromParts(yearText: string, monthText: string): MonthlySummaryPeriod | undefined {
  const year = Number(yearText);
  const month = Number(monthText);
  if (!Number.isInteger(year) || !Number.isInteger(month) || month < 1 || month > 12) {
    return undefined;
  }
  return { year, month };
}
